src/python/dxpy/sugar/transfers.py

```diff
@@ -0,0 +1,952 @@
+"""
+This module contains utilities for file upload and download. File downloads can be
+done in blocking or non-blocking mode.
+
+This module also provides the Uploader and Downloader classes for managing multiple
+concurrent up/downloads. Individual files, lists of files, and dicts of files can
+be queued for up/download with different settings. Each enqueue event is associated
+with a name. Calling the `wait()` method blocks until all up/downloads have completed
+and returns a dict mapping names to either local paths (for Downloader) or dxlinks
+(for Uploader).
+
+```
+with Downloader() as downloader:
+  downloader.enqueue_file("reference_tgz", "file-XXX")
+  downloader.enqueue_list("fastqs", ["file-YYY", "file-ZZZ"], skip_decompress=True)
+  local_files = downloader.wait()
+
+subprocess.call(
+    "bwa mem {reference_tgz} {fastqs}".format(**local_files),
+    output_filename="output.bam"
+)
+
+with Uploader() as uploader:
+  uploader.enqueue_file("output_bam", "output.bam", skip_compress=True)
+  uploader.enqueue_file("output_bai", "output.bam.bai", skip_compress=True)
+  return uploader.wait()
+```
+
+TODO: switch to using dxpy vs calls to dx whenever possible
+"""
+
+# Copyright (C) 2013-2019 DNAnexus, Inc.
+#
+# This file is part of dx-toolkit (DNAnexus platform client libraries).
+#
+#   Licensed under the Apache License, Version 2.0 (the "License"); you may not
+#   use this file except in compliance with the License. You may obtain a copy
+#   of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#   WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#   License for the specific language governing permissions and limitations
+#   under the License.
+from __future__ import print_function, unicode_literals, division, absolute_import
+import concurrent.futures
+import contextlib
+import copy
+import logging
+import multiprocessing
+import os
+import re
+import sys
+
+import dxpy
+from dxpy.compat import basestring, makedirs
+from dxpy.utils.resolver import data_obj_pattern
+from dxpy.sugar import processing as proc
+
+
+# Helpers - these belong in other modules
+
+import tempfile
+
+
+@contextlib.contextmanager
+def tmpfile(*args, **kwargs):
+    """
+    Create a temporary file, yield it, and delete it before returning.
+
+    Yields:
+        A path to a temporary file.
+
+    Notes:
+        This method is needed distinct from :class:`tempfile.TemporaryFile` in the
+        case where python needs to write to the file and then a subprocess needs
+        to read from the file.
+    """
+    path = tempfile.mkstemp(*args, **kwargs)[1]
+    try:
+        yield path
+    finally:
+        if os.path.exists(path):
+            os.remove(path)
```

Comment:
Note: This should be used from the context manager module, I am assuming it is only here temporarily and will be removed once the context managers module is merged in. In that case, I think it would be better to put it in a separate temporary context manager module in this PR and import it from there.

Reply:
Depends on whether you decide to include it in context.py or if it should remain private to this module. If you want me to keep it private, I'll prefix the function name with "_".

Reply:
I think a better implementation for this would be to use `tempfile.TemporaryNamedFile`, I'll add a comment below with what I mean.